src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -1096,12 +1098,41 @@ internal void RemoveAdditionalDocument(IVisualStudioHostDocument document)
             lock (_gate)
             {
                 _additionalDocuments.Remove(document.Id);
-                _documentMonikers.Remove(document.Key.Moniker);
+                RemoveMoniker(document);
             }
 
             UninitializeAdditionalDocument(document);
         }
 
+        private void AddMoniker(IVisualStudioHostDocument document)
+        {
+            var moniker = document.Key.Moniker;
+            if (_documentMonikers.TryGetValue(moniker, out var value))
+            {
+                _documentMonikers[moniker] = (value.document, value.refCount++);
+            }
+            else
+            {
+                _documentMonikers.Add(moniker, (document, 1));
+            }
+        }
+
+        private void RemoveMoniker(IVisualStudioHostDocument document)
+        {
+            var moniker = document.Key.Moniker;
+            var value = _documentMonikers[moniker];
```

Comment:
Good catch, the code previously just had `_documentMonikers.Remove()`, which wouldn't throw.  I'll guard.